internal/pkg/aws/ec2/ec2.go

```diff
@@ -377,3 +364,76 @@ func toEC2Filter(filters []Filter) []*ec2.Filter {
 	}
 	return ec2Filter
 }
+
+type routeTable ec2.RouteTable
+
+// IsMain returns true if the route table is the default route table for the VPC.
+// If a subnet is not associated with a particular route table, then it will default to the main route table.
+func (rt *routeTable) IsMain() bool {
+	for _, association := range rt.Associations {
+		if aws.BoolValue(association.Main) {
+			return true
+		}
+	}
+	return false
+}
+
+// HasIGW returns true if the route table has a route to an internet gateway.
+func (rt *routeTable) HasIGW() bool {
+	for _, route := range rt.Routes {
+		if strings.HasPrefix(aws.StringValue(route.GatewayId), internetGatewayIDPrefix) {
+			return true
+		}
+	}
+	return false
+}
+
+// AssociatedSubnets returns the list of subnet IDs associated with the route table.
+func (rt *routeTable) AssociatedSubnets() []string {
+	var subnetIDs []string
+	for _, association := range rt.Associations {
+		if association.SubnetId == nil {
+			continue
+		}
+		subnetIDs = append(subnetIDs, aws.StringValue(association.SubnetId))
+	}
+	return subnetIDs
+}
+
+// routeTableIndex holds cached data to quickly return information about route tables in a VPC.
+type routeTableIndex struct {
+	// Route table that subnets default to. There is always one main table in the VPC.
+	mainTable *routeTable
+
+	// Explicit route table association for a subnet. A subnet can only be associated to one route table.
+	routeTableForSubnet map[string]*routeTable
+}
+
+func indexRouteTables(tables []*ec2.RouteTable) *routeTableIndex {
+	index := &routeTableIndex{
+		routeTableForSubnet: make(map[string]*routeTable),
+	}
+	for _, table := range tables { // Index all properties in a single pass.
+		table := (*routeTable)(table)
```

Comment:
do we need to check the type in case the type can't be asserted?

Reply:
no need boss it's a compile type check